ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/snap/SnapServer.java

```diff
@@ -14,66 +14,616 @@
+    this.protocolContext
+        .flatMap(ProtocolContext::getSynchronizer)
+        .filter(z -> z instanceof DefaultSynchronizer)
+        .map(DefaultSynchronizer.class::cast)
+        .ifPresentOrElse(
+            z -> this.listenerId.set(z.subscribeInitialSync(this)),
+            () -> LOGGER.warn("SnapServer created without reference to sync status"));
+  }
+
+  /**
+   * Create a snap server without registering a listener for worldstate initial sync events or
+   * priming worldstates by root hash. Used by unit tests.
+   */
+  @VisibleForTesting
+  SnapServer(
+      final EthMessages snapMessages,
+      final WorldStateStorageCoordinator worldStateStorageCoordinator,
+      final Function<Hash, Optional<BonsaiWorldStateKeyValueStorage>> worldStateStorageProvider) {
+    this.snapServerEnabled = true;
+    this.snapMessages = snapMessages;
+    this.worldStateStorageCoordinator = worldStateStorageCoordinator;
+    this.worldStateStorageProvider = worldStateStorageProvider;
+    this.protocolContext = Optional.empty();
+  }
+
+  @Override
+  public void onInitialSyncCompleted() {
+    start();
+  }
+
+  @Override
+  public void onInitialSyncRestart() {
+    stop();
+  }
+
+  public synchronized SnapServer start() {
+    if (!isStarted.get() && snapServerEnabled) {
+      // if we are bonsai and full flat, we can provide a worldstate storage:
+      var worldStateKeyValueStorage = worldStateStorageCoordinator.worldStateKeyValueStorage();
+      if (worldStateKeyValueStorage.getDataStorageFormat().equals(DataStorageFormat.BONSAI)
+          && worldStateStorageCoordinator.isMatchingFlatMode(FlatDbMode.FULL)) {
+        LOGGER.debug("Starting snap server with Bonsai full flat db");
+        var bonsaiArchive =
+            protocolContext
+                .map(ProtocolContext::getWorldStateArchive)
+                .map(BonsaiWorldStateProvider.class::cast);
+        var cachedStorageManagerOpt =
+            bonsaiArchive.map(archive -> archive.getCachedWorldStorageManager());
+
+        if (cachedStorageManagerOpt.isPresent()) {
+          var cachedStorageManager = cachedStorageManagerOpt.get();
+          this.worldStateStorageProvider = cachedStorageManager::getStorageByRootHash;
+
+          // when we start we need to build the cache of latest 128 worldstates
+          // trielogs-to-root-hash:
+          var blockchain = protocolContext.map(ProtocolContext::getBlockchain).orElse(null);
+
+          // at startup, prime the latest worldstates by roothash:
+          cachedStorageManager.primeRootToBlockHashCache(blockchain, PRIME_STATE_ROOT_CACHE_LIMIT);
+
+          var flatDbStrategy =
+              ((BonsaiWorldStateKeyValueStorage)
+                      worldStateStorageCoordinator.worldStateKeyValueStorage())
+                  .getFlatDbStrategy();
+          if (!flatDbStrategy.isCodeByCodeHash()) {
+            LOGGER.warn("SnapServer requires code stored by codehash, but it is not enabled");
```

Comment:
what we are doing in this case ? not sure to understand if we are stopping the snapserver ?

Reply:
for now, snap server will operate as normal with just a warning.  Essentially we will just serve empty code requests if the storage hasn't migrated.  If we think it is necessary we can refuse to start snap server if we are not in code-by-code hash storage mode.  I was thinking that we would address this corner case via the storage migration subcommand PR